cal/cal/static/js/services.js

```diff
@@ -44,53 +43,83 @@ analyticsApp.service("TagService", ['$http', '$q', function($http, $q) {
   this.tags = {};
 
   this.getTags = function(filterKey, start, end, calendarIds) {
+    var calendarTags = {}; // {calendarId : list of tags}
+    var accumulatedTags = {};
+    var requests = [];
+    var calendarRequestsMap = {}; // {request index : calendarId}
+    var tags = [];
 
     if (!filterKey) {
       throw "filterKey must always be supplied";
     }
 
     if (start || end) {
       // If the start and end time match the given filterKey
-      var keyFromParameters = start.toISOString() + " " + end.toISOString() +
-        " " + calendarIds.join(' ');
+      var keyFromParameters = start.toISOString() + " " + end.toISOString();
       if (filterKey !== keyFromParameters) {
         throw "filterKey doesn't match given start and end times";
       }
     }
-    // Attempt to return cached tags
-    if (_this.tags[filterKey]) {
-      return $q.when(_this.tags[filterKey]);
+
+    for (var i = 0; i < calendarIds.length; i++) {
+      var cacheKey = calendarIds[i] + " " + filterKey;
+      if (this.tags[cacheKey]) {
+        calendarTags[calendarIds[i]] = this.tags[cacheKey];
+      } else {
+        requests.push($http({
+          method: 'GET',
+          url: '/v1/tags.json',
+          cache: true,
+          params: {
+            start: (start)? start.toISOString() : null,
+            end: (end)? end.toISOString() : null,
+            calendar_ids: JSON.stringify([calendarIds[i]]),
+            timezone: moment.tz.guess()
+          }
+        }));
+        calendarRequestsMap[i] = calendarIds[i];
+      }
     }
 
-    // Request the tags and return a promise
-    return $http({
-      method: 'GET',
-      url: '/v1/tags.json',
-      cache: true,
-      params: {
-        start: (start)? start.toISOString() : null,
-        end: (end)? end.toISOString() : null,
-        calendar_ids: JSON.stringify(calendarIds),
-        timezone: moment.tz.guess()
+    return $q.all(requests).then(function(responses) {
+      for (var j = 0; j < responses.length; j++) {
+        var data = responses[j].data.results;
+        var calId = calendarRequestsMap[j];
+        var cacheKey = calId + " " + filterKey;
+        calendarTags[calId] = data;
+        _this.tags[cacheKey] = data;
       }
-    }).then(function successCallback(response) {
-      _this.tags[filterKey] = [];
-      for (var i = 0; i < response.data.results.length; i++) {
-        var tag = response.data.results[i];
-        _this.tags[filterKey].push({
-          id: tag.id,
-          label: tag.label,
-          keywords: tag.keywords,
-          hours: tag.hours
-        });
+
+      for (var calendar in calendarTags) {
+        for (var n = 0; n < calendarTags[calendar].length; n++) {
+          var tag = calendarTags[calendar][n];
+          if (accumulatedTags.hasOwnProperty(tag.id)) {
```

Comment:
[hasOwnProperty should take in a string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty)